DigitalLearningSolutions.Web/Services/EmailVerificationService.cs

```diff
@@ -0,0 +1,115 @@
+﻿namespace DigitalLearningSolutions.Web.Services
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.Models;
+    using DigitalLearningSolutions.Data.Models.Email;
+    using DigitalLearningSolutions.Data.Models.User;
+    using DigitalLearningSolutions.Data.Utilities;
+    using MimeKit;
+
+    public interface IEmailVerificationService
+    {
+        bool AccountEmailIsVerifiedForUser(int userId, string email);
+
+        void SendVerificationEmails(
+            UserAccount userAccount,
+            IEnumerable<PossibleEmailUpdate> unverifiedEmails,
+            string baseUrl
+        );
+    }
+
+    public class EmailVerificationService : IEmailVerificationService
+    {
+        private readonly IEmailVerificationDataService emailVerificationDataService;
+        private readonly IEmailService emailService;
+        private readonly IClockUtility clockUtility;
+
+        public EmailVerificationService(
+            IEmailVerificationDataService emailVerificationDataService,
+            IEmailService emailService,
+            IClockUtility clockUtility
+        )
+        {
+            this.emailVerificationDataService = emailVerificationDataService;
+            this.emailService = emailService;
+            this.clockUtility = clockUtility;
+        }
+
+        public bool AccountEmailIsVerifiedForUser(int userId, string email)
+        {
+            return emailVerificationDataService.AccountEmailIsVerifiedForUser(userId, email);
+        }
+
+        public void SendVerificationEmails(
+            UserAccount userAccount,
+            IEnumerable<PossibleEmailUpdate> unverifiedEmails,
+            string baseUrl
+        )
+        {
+            foreach (var emailGroup in unverifiedEmails.GroupBy(emailUpdate => emailUpdate.NewEmail))
+            {
+                var hash = Guid.NewGuid().ToString();
+                var hashId = emailVerificationDataService.CreateEmailVerificationHash(hash, clockUtility.UtcNow);
+
+                foreach (var emailUpdate in emailGroup)
+                {
+                    UpdateEmailVerificationHashId(userAccount.Id, emailUpdate.CentreId, hashId);
+                }
+
+                emailService.SendEmail(
+                    GenerateVerificationEmail(userAccount, hash, emailGroup.Key!, baseUrl)
+                );
+            }

```

Comment:
I have a suggestion in another comment which should allow you to simplify this method to
```
            foreach (var emailGroup in unverifiedEmails.GroupBy(emailUpdate => emailUpdate.NewEmail))
            {
                var hash = Guid.NewGuid().ToString();
                var hashId = emailVerificationDataService.CreateEmailVerificationHash(hash, clockUtility.UtcNow);
                var emailAddress = emailGroup.Key!;

                UpdateEmailVerificationHashId(userAccount.Id, emailAddress, hashId);

                emailService.SendEmail(
                    GenerateVerificationEmail(userAccount, hash, emailAddress, baseUrl)
                );
            }
```